build/patches/Remove-http-referrals-in-cross-origin-navigation.patch

```diff
@@ -0,0 +1,103 @@
+                             : R.string.text_off);
+         }
+ 
++        ChromeSwitchPreference enableFeferrerPref = findPreference(PREF_ENABLE_REFERRERS);
++        enableFeferrerPref.setChecked(prefService.getBoolean(Pref.ENABLE_REFERRERS));
++        enableFeferrerPref.setOnPreferenceChangeListener(this);
++
+         Preference preloadPagesPreference = findPreference(PREF_PRELOAD_PAGES);
+         if (preloadPagesPreference != null) {
+             preloadPagesPreference.setSummary(
+diff --git a/chrome/browser/ui/android/strings/android_chrome_strings.grd b/chrome/browser/ui/android/strings/android_chrome_strings.grd
+--- a/chrome/browser/ui/android/strings/android_chrome_strings.grd
++++ b/chrome/browser/ui/android/strings/android_chrome_strings.grd
+@@ -873,6 +873,9 @@ CHAR_LIMIT guidelines:
+ 
+ For example, some websites may respond to this request by showing you ads that aren’t based on other websites you’ve visited. Many websites will still collect and use your browsing data — for example to improve security, to provide content, ads and recommendations, and to generate reporting statistics.
+       </message>
++      <message name="IDS_ENABLE_REFERRERS_TITLE" desc="Title for 'Enable referrers' preference">
++        Enable http referrer header
++      </message>
+       <message name="IDS_CAN_MAKE_PAYMENT_TITLE" desc="Title for preference to allow websites to know whether you have payment methods available through PaymentRequest.CanMakePayment interface">
+         Access payment methods
+       </message>
+diff --git a/content/browser/renderer_host/navigation_request.cc b/content/browser/renderer_host/navigation_request.cc
+--- a/content/browser/renderer_host/navigation_request.cc
++++ b/content/browser/renderer_host/navigation_request.cc
+@@ -425,6 +425,13 @@ void AddAdditionalRequestHeaders(
+         blink::mojom::Referrer(GURL(), network::mojom::ReferrerPolicy::kNever);
+   }
+ 
++  if (!url::IsSameOriginWith(referrer->url.GetAsReferrer(), url) &&
++      frame_tree_node->IsOutermostMainFrame()) {
++    // remove referrer if the navigation is done on the top frame
++    *referrer =
++        blink::mojom::Referrer(GURL(), network::mojom::ReferrerPolicy::kNever);
++  }
++
+   // Next, set the HTTP Origin if needed.
+   if (NeedsHTTPOrigin(headers, method)) {
+     url::Origin origin_header_value = initiator_origin.value_or(url::Origin());
+diff --git a/services/network/public/cpp/resource_request.h b/services/network/public/cpp/resource_request.h
+--- a/services/network/public/cpp/resource_request.h
++++ b/services/network/public/cpp/resource_request.h
+@@ -130,7 +130,7 @@ struct COMPONENT_EXPORT(NETWORK_CPP_BASE) ResourceRequest {
+   std::vector<GURL> navigation_redirect_chain;
+ 
+   GURL referrer;
+-  net::ReferrerPolicy referrer_policy = net::ReferrerPolicy::NEVER_CLEAR;
++  net::ReferrerPolicy referrer_policy = net::ReferrerPolicy::REDUCE_GRANULARITY_ON_TRANSITION_CROSS_ORIGIN;
```

Comment:
this i think is a chromium bug, because that policy does not match the default and is used as is in service workers

Reply:
It seems a quite big bug?

Reply:
I don't know what to tell you. I noticed it by logging in [URLRequestJob::ComputeReferrerForPolicy](https://source.chromium.org/chromium/chromium/src/+/main:net/url_request/url_request_job.cc;l=290;drc=f97e7e130b02a0fee5a06aa9cdf25d3a0a3715d0;bpv=1;bpt=1)

I noticed that the chromium team has to (always) follow the specs, maybe the specs say nothing about it (I did not check). in that case they often change the specifications first, then the code, maybe this is going on.

Reply:
Perhaps there is something else going on that makes this bug less important; have you tried submitting it upstream as a security issue? Maybe they will mark yours as a duplicate of the main one.

Reply:
> have you tried submitting it upstream as a security issue?

no, maybe I'll try, but I wouldn't classify it as a security bug, after all, sites can choose how to manage referrals